rust/src/feed.rs

```diff
@@ -0,0 +1,347 @@
+use byteorder::{BigEndian, ByteOrder};
+use hex::{FromHex, ToHex};
+use serde::{Deserialize, Serialize};
+use std::fmt;
+use std::str::FromStr;
+
+/// Represents the feed report schema version.
+///
+/// The `FeedVersion` struct wraps a `u16` integer representing the version
+/// of the feed report schema.
+///
+/// # Examples
+///
+/// ```rust
+/// use data_streams_sdk::feed::FeedVersion;
+///
+/// let version = FeedVersion(1);
+/// println!("Feed version: {}", version.0);
+/// ```
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct FeedVersion(pub u16);
+
+/// Represents a 32-byte identifier.
+///
+/// The `ID` struct encapsulates a 32-byte array and provides methods for
+/// parsing from and converting to hexadecimal strings, as well as extracting
+/// the feed version from the identifier.
+///
+/// # Examples
+///
+/// ```rust
+/// use data_streams_sdk::feed::ID;
+///
+/// let id = ID::from_hex_str("0x00016b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472").unwrap();
+/// println!("ID: {}", id);
+/// ```
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct ID([u8; 32]);
+
+impl ID {
+    /// Parses an `ID` from a hexadecimal string with a "0x" prefix.
+    ///
+    /// # Arguments
+    ///
+    /// * `s` - A string slice that holds the hexadecimal representation of the ID.
+    ///
+    /// # Returns
+    ///
+    /// * `Ok(ID)` if parsing is successful.
+    /// * `Err(&'static str)` if the input string is invalid.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if:
+    /// - The string does not start with "0x" or "0X".
+    /// - The string length after the prefix is not exactly 64 characters (32 bytes).
+    /// - The string contains invalid hexadecimal characters.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use data_streams_sdk::feed::ID;
+    ///
+    /// let id = ID::from_hex_str("0x00016b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472").unwrap();
+    /// ```
+    pub fn from_hex_str(s: &str) -> Result<Self, &'static str> {
```

Comment:
`thiserror` would be ideal, yes :). Ultimately, you really want whatever's on the `E` side of `Result` to be a type that implements `Error`, otherwise the consumers of the library are going to have a more difficult time handling the value.